src/main/java/build/buildfarm/worker/InputFetchStage.java

```diff
@@ -143,12 +200,27 @@ private void linkInputs(
     for (DirectoryNode directoryNode : directory.getDirectoriesList()) {
       Digest digest = directoryNode.getDigest();
       String name = directoryNode.getName();
+      OutputDirectory childOutputDirectory = outputDirectory != null
+          ? outputDirectory.directories.get(name) : null;
       Path dirPath = execDir.resolve(name);
-      Files.createDirectories(dirPath);
-      linkInputs(dirPath, digest, directoriesIndex, inputFiles);
+      if (childOutputDirectory != null || !worker.config.getLinkInputDirectories()) {
+        Files.createDirectories(dirPath);
+        linkInputs(dirPath, digest, directoriesIndex, childOutputDirectory, inputFiles, inputDirectories);
+      } else {
+        inputDirectories.add(digest);
+        linkDirectory(dirPath, digest, directoriesIndex);
+      }
     }
   }
 
+  private void linkDirectory(
```

Comment:
Do I understand correctly that you are symlinking only the direct children of the root directory, in full, never recursively? I mean, if input of action A is:
a/b/c/d/foo.txt
and input of action B is:
a/x/b/c/d/foo.txt
then there will be no reuse of the b/c/d/foo.txt subtree? If so, is that because you think it would not be as frequently useful?

Reply:
You are correct, and it is not that the net result of recursive subtrees would be unuseful, but that none of my (freely available open-source versions of) tools were able to function with multiple levels of symlink subdirectories (too much realpath() usage). Even if the base resolution of libraries worked for the loader, runtimes would typically fail, and the top-level directory based on how bazel functions and the nature of the highest level yielded strong performance improvements (since only an intersection with outputs will prevent symlinking).